### DISCUSSION TITLE
nil inputs in service calls

### DISCUSSION THREAD

**Discussion by siggimoo** [aws/aws-sdk-go-v2#1911] [Q&A]
I just happened to notice while reviewing the source code that passing a `nil` input to service calls seems to be allowed, with a zero-value struct being used in its place. For instance, the following two statements are effectively identical:
```
output, err := s3Service.ListBuckets(ctx, nil)
output, err := s3Service.ListBuckets(ctx, &s3.ListBucketsInput{})
```
While this is handy, I can't seem to find the behavior documented anywhere and I don't want to rely on it if it's not a standard convention. Does anyone know anything about this?

**Comment by RanVaknin** [comment]
Hi @siggimoo ,

I'm not sure what exactly your question is. 
We allow nil inputs because the request code handles nil.

```go
func (c *S3) ListBucketsRequest(input *ListBucketsInput) (req *request.Request, output *ListBucketsOutput) {
	op := &request.Operation{
		Name:       opListBuckets,
		HTTPMethod: "GET",
		HTTPPath:   "/",
	}

// You can see here that nil inputs get converted to an empty struct.
	if input == nil {
		input = &ListBucketsInput{}
	}

	output = &ListBucketsOutput{}
	req = c.newRequest(op, input, output)
	return
}
```

Thanks,
Ran~

**Comment by siggimoo** [comment]
I am aware that at least some of the code guards against nil inputs -- I haven't checked every single service call. What I'm not readily able to find is any mention of this in the documentation (e.g. [here](https://aws.github.io/aws-sdk-go-v2/docs/making-requests/#calling-service-operations)). So it's unclear whether this is behavior I can trust to always be there.

**Reply by RanVaknin** [reply]
Yes I'd say that it would be the case for 99% of the times since the SDK clients are auto-generated aside from a handful of hand-written ones.
